nn/nntree.py

```diff
@@ -0,0 +1,489 @@
+    output from an IceParsing neural network model and working with the
+    result.
+
+"""
+
+from __future__ import print_function
+from enum import IntEnum
+import logging
+import inspect
+
+import tokenizer
+from reynir import bintokenizer
+from reynir import matcher
+
+
+class GRAMMAR:
+    CASES = {"nf", "þf", "þgf", "ef"}
+    GENDERS = {"kk", "kvk", "hk"}
+    NUMBERS = {"et", "ft"}
+    PERSONS = {"p1", "p2", "p3"}
+
+    TENSE = {"þt", "nt"}
+    DEGREE = {"mst", "esb", "evb"}  # fst
+
+    VOICE = {"mm", "gm"}
+    MOOD = {"fh", "lhþt", "lhnt", "vh", "bh"}
+
+    MISC = {"sagnb", "subj", "abbrev", "op", "none"}
+
+
+class KEY:
+    long_terminal = "a"
+    bin_variants = "b"
+    bin_category = "c"
+    bin_fl = "f"
+    nonterminal_tag = "i"  # nonterminal
+    token_index = "ix"
+    kind = "k"  # token or nonterminal kind
+    nonterminal_name = "n"
+    lemma = "s"
+    short_terminal = "t"  # matchable categories
+    text = "x"
+    children = "p"
+
+
+FieldKeyToName = {
+    value: key for (key, value) in inspect.getmembers(KEY) if not key.startswith("_")
+}
+
+
+def flat_is_nonterminal(string):
+    return string.isupper() and "_" not in string
+
+
+def flat_is_terminal(string):
+    return string.islower()
+
+
+def flat_is_left_nonterminal(string):
+    return flat_is_nonterminal(string) and "/" not in string
+
+
+def flat_is_right_nonterminal(string):
+    return flat_is_nonterminal(string) and "/" in string
+
+
+def flat_matching_nonterminal(string):
+    if "/" in string:
+        return string[1:]
+    return "/" + string
+
+
+class Node:
+    """ Generic tree implementation """
+
+    def __init__(self, name, is_terminal=False, data=None):
+        self.name = name
+        self.data = data
+        self.children = []
+        self.is_terminal = is_terminal
+
+    def add_child(self, node):
+        self.children.append(node)
+        return self
+
+    def has_children(self):
+        return bool(self.children)
+
+    def to_dict(self):
+        if not self.children:
+            json_node = _json_terminal_node(self.name, self.data)
+        else:
+            json_node = _json_nonterminal_node(self.name)
+            json_node[KEY.children] = [c.to_dict() for c in self.children]
+        return json_node
+
+    def to_simple_tree(self):
+        return matcher.SimpleTree([[self.to_dict()]])
+
+    def to_postfix(self):
+        """ Export tree to postfix ordering
+            with node-labels as keys """
+        result = []
+        for child in self.children:
+            result.extend(child.to_postfix())
+        result.append(self.name)
+        return result
+
+    def __str__(self):
+        text = (" " + self.data) if self.data is not None else ""
+        return "<" + self.name + text + ">"
+
+    def __repr__(self):
+        if self.is_terminal:
+            return self.__str__()
+        result = [self.__str__()]
+        for child in self.children:
+            result.extend(child.__repr__())
```

Comment:
Why `.extend()`? Why not `.append()`?

Reply:
Extend is to flatten the tree, append would nest lists.

Reply:
```
>>>> r = []
>>>> r.extend("Some string")
>>>> r
['S', 'o', 'm', 'e', ' ', 's', 't', 'r', 'i', 'n', 'g']
>>>>
```

Reply:
__repr__() always returns a string, not a list.